src/schema/builder.rs

```diff
@@ -0,0 +1,592 @@
+use failure::Fail;
+use std::collections::HashSet;
+
+use super::*;
+use crate::schema::data::*;
+
+#[derive(Fail, Debug)]
+#[fail(display = "Invalid schema construction: {}", _0)]
+pub struct BuilderError(pub String);
```

Comment:
nit: perhaps `SchemaBuilderError` given that the struct is called SchemaBuilder?

Reply:
Yep :P

Reply:
thanks!